File: section-03-basic-maths/11-count-of-prime-numbers-till-n/count-of-prime-numbers-till-n-opt.cs
class Solution
{
    public bool isPrime(int n)
    {
        if (n < 2)
            return false;
        for (int i = 2; i * i <= n; i++)
        {
            if (n % i == 0)
                return false;
        }
        return true;
    }

    public int primeUptoN(int n)
    {
        int cnt = 0;
        for (int i = 2; i <= n; i++)
        {
            if (this.isPrime(i))
                cnt++;
        }
        return cnt;
    }
}

/* 
Time Complexity
O(n * sqrt(n)) where n is the input integer; the outer loop runs from 2 to n, and the inner loop (isPrime) runs up to sqrt(i) for each i.
Space Complexity
O(1) as only a constant amount of extra space is used for variables (cnt, i) regardless of the input size.
 */

File: section-03-basic-maths/11-count-of-prime-numbers-till-n/count-of-prime-numbers-till-n-opt.py
class Solution:
    def isPrime(self, n):
        if n < 2:
            return False
        i = 2
        while i * i <= n:
            if n % i == 0:
                return False
            i += 1
        return True

    def primeUptoN(self, n):
        counter = 0
        i = 2
        while i <= n:
            if self.isPrime(i):
                counter += 1
            i += 1
        return counter
